### DISCUSSION TITLE
Complexity of running the tests locally

### DISCUSSION THREAD

**Discussion by avelino** [prest/prest#562] [API Server] [answered]
The `postgres` adapter depends on the postgreSQL (database) server to run (`go test`), i.e. it was developed with integration testing.

Integration testing is necessary, but it raises the barrier for new members to contribute, i.e. for a person to make any contribution he needs to learn to configure the development environment and only then contribute. I think it's great to have people immersed in the project, but this is not the reality of open source projects that people contribute in their spare time (usually without pay)

I don't think it is necessary to "remove" the integration test but put it in its place:
1. unit test that does not depend on external services (e.g. PostgreSQL) running with `go test ....`, so our people were able to run the tests locally (on their machines)
2. Integration test: run only on CI (**GitHub Actions** or Travis CI) on pull request - doesn't necessarily have to be written with `go test`

What would be the best option to write `postgres` adapter **mock**?

Ref:
- #510
- #555 
- #558 
- #472 
- … as we see above we have several issues related to this subject - sad

**Comment by felipeweb** [comment]
we can use https://github.com/zombiezen/postgrestest to be able to run tests easier

**Reply by DiegoSantosWS** [reply]
> we can use https://github.com/zombiezen/postgrestest to be able to run tests easier

Thanks @felipeweb

**Comment by DiegoSantosWS** [comment]
Hi People, I'm analyzing an possible refactoring on code to created new Unit test without heve a real connection with databases.

But, on my opinion, it's best applicate a refactor in interface `Adapters`. Bucause to executed tests using the [Pg test](https://github.com/zombiezen/postgrestest) is important realize an change to connection with database. because many function require the `connection.Get`.

Let's talk about the reformulation?

**Reply by avelino** [reply]
What would be your proposal for change?

I agree that we need to remove the e2e test that exists today (dependency on postgresql to run the tests).

Your proposal is to use the lib [postgrestest](https://github.com/zombiezen/postgrestest)?

**Reply by DiegoSantosWS** [reply]
> What would be your proposal for change?

Create de connection with database using interfaces

```go
type ConnGetter interface {
     Get() ( DB *sqlx.DB, error, err)
}

func Load(conn ConnGetter) {
    db, err := conn.Get()
}
```

with this example, possibility creating the mock.

> I agree that we need to remove the e2e test that exists today (dependency on postgresql to run the tests).
> Your proposal is to use the lib [postgrestest](https://github.com/zombiezen/postgrestest)?

Yeah, use the postgrestest.

**Reply by avelino** [reply]
@DiegoSantosWS I like the idea of working with mock, I believe that postgrest will meet our needs

can you work on this change?

**Reply by DiegoSantosWS** [reply]
> @DiegoSantosWS I like the idea of working with mock, I believe that postgrest will meet our needs
> 
> can you work on this change?

Yeah, i followed with this change, do you open the issue?

**Reply by avelino** [reply]
> Yeah, i followed with this change, do you open the issue?

@DiegoSantosWS here https://github.com/prest/prest/issues/579

**Comment by avelino** [comment, marked as answer]
I tried to simplify as much as possible the way to run the tests locally using docker (to not keep the dependency on having local postgres)

we haven't done a mock migration of postgres yet, but I believe it is the best way forward
https://github.com/prest/prest#running-local-unit-tests-integratione2e
